### ISSUE TITLE
do_eval is broken with args.load_target_train_checkpoint

### ISSUE THREAD

**Issue by sleepinyourhat** [nyu-mll/jiant#741]
We only save the special model_state_untrained_pre_target_train.th checkpoint if there is no load_target_train_checkpoint, but we look for it regardless at the end of the target task training section in main.

save: https://github.com/nyu-mll/jiant/blob/master/main.py#L142

load: https://github.com/nyu-mll/jiant/blob/master/main.py#L525

I can't really figure out what the code around line 525 is meant to do, so I'm not sure what direction is best for a fix. @pruksmhc, @W4ngatang – does this look familiar to either of you?

**Comment by pruksmhc** [comment]
The code around 525 is because for finetuning each task, we want to start off with a fresh copy of BERT. If we pretrain first, the fresh copy would be the weights of the best performing BERT weights in pretraining stage, else it would be model_state_untrained_pre_target_train.th that we save previously.

**Comment by sleepinyourhat** [comment]
Ah. We're doing it one extra time, then, since we do it after every task, even the last. We should fix the path finding and also fix that, since each save/load is a ton of file i/o. Do you have bandwidth to take this on?

**Comment by pruksmhc** [comment]
I'm focusing on winogender and a few other things for the first part of today, and can take this on after. Feel free to take this one if you have time before then!

**Comment by sleepinyourhat** [comment]
Sounds good.

**Comment by sleepinyourhat** [comment]
I took a look at this and it's bigger than I can really take on today/tomorrow. I think we'd need to:
- Add some logic at the start of the pretraining section to identify what checkpoint to load.
- Save model_state_untrained_pre_target_train.th iff no checkpoint is available.
- Add logic to decide when we actually need to reload the checkpoint.
- Rewrite/test the eval checkpoint loading system to ensure that it still properly supports

**Comment by sleepinyourhat** [comment]
Sorry!

**Comment by pruksmhc** [comment]
What is the thinking behind the checking logic at the start of the pretraining section? Don't we always want to train from a fresh start? (unless we have something like load_pretrain_checkpoint).

**Comment by sleepinyourhat** [comment]
The idea is that most of the code/logic only needs to happen once:
- Deciding whether to save a model_state_untrained_pre_target_train.th checkpoint.
- Deciding what checkpoint (if any) to load at the start of target task training.
- Deciding whether we need to reload that checkpoint for future tasks. (I think this is basically just checking transfer_paradigm == finetune.)

But yeah, at the start of each task, we'll have to have a conditional that checks if we need to actually do a torch.load.

**Comment by pruksmhc** [comment]
Talking about adding logic about reloading a checkpoint, I also don't think we need task_names_to_avoid_loading, since in transfer_pardigm=frozen, after we train the model, we then allow loading of weights relating to that task from the pre-target train checkpoint. That doesn't make sense to me since we don't use it again (since we've already target trained on that task). Since it seems like we're keeping target_train to just training for each task individually, we only train on each task once through.

**Comment by sleepinyourhat** [comment]
My understanding of that mechanism is that it related to the fact that we don't have a good restore function.

The idea is that if we partially do fine-tuning, then training crashes, we don't want to accidentally load the fine-tuned model parameters but then start training at step 0, since that'll give misleading results. (Like strangely good accuracy at step 0.)

If we're already doing other things to avoid that situation, then we're good.

**Comment by pruksmhc** [comment]
Yeah, I think currently, the only way that you can load the fine-tuned model parameters in target training stage is if you explicitly use load_target_train_checkpoint, so shouldn't be able to "accidentally" load finetuned weights easily now (if you even can).

**Comment by sleepinyourhat** [comment]
Cool. Does this mean that frozen and non-frozen now behave the same way (i.e., one file per task)? 

That wastes some disk space (for ELMo, we'd be saving ELMo once for each task), but it seems like the right thing to do for now to avoid wonky code complexity.

**Comment by pruksmhc** [comment]
Yep!
